File: PythonLeet/ValidParantheses/valid.py
class Solution(object):
    
    def isValid(s):
        closed = {"{": "}", "(": ")", "[": "]"}
        
        stack = Stack()
        stack.push(s[0])
        i = 1
        
        while(stack.isEmpty() == False):
            if self.isOpen(s[i]):
                stack.push(closed[s[i]])
                i += 1
            else:
                curr = stack.pop()
            # elif self.isClosed(s[i]):
            #     curr = stack.pop()
            #     if s[i] != curr:
            #         return False
            #     else:
            #         i += 1
        return True
        
        
    
    def isOpen(self, c):
        open_chars = ["(", "[", "{"]
        if c in open_chars:
            return True
        else:
            return False
    
    def isClosed(self, c):
        closed_chars = ["}", ")", "]"]
        if c in closed_chars:
            return True
        else:
            return False

def main():
    s = Solution()
    print(s.isValid("()"))

File: PythonLeet/ValidParantheses/valid.java
class Solution {

    public boolean isValid(String s) {
        if(s.length() == 0){
            return true;
        }
        //parantheses = {"}": "{", "]": "[", ")": "("
        HashMap<Character, Character> paran = new HashMap<Character, Character>();
        paran.put('{', '}');
        paran.put('[',']' );
        paran.put('(', ')');

        Stack<Character> stack = new Stack<Character>();
        for(int i = 0; i < s.length(); i++){
            Character curr = s.charAt(i);
            if(isOpen(curr)){
                stack.push(paran.get(curr));
            }
            else if(isClosed(curr)){
                if (stack.isEmpty()){
                    return false;
                }
                else {
                    Character top = stack.pop();
                    if (curr != top){
                        return false;
                    }
                }
            }
            //System.out.println(stack.peek());
        }
        if (stack.isEmpty()){
            return true;
        }
        return false;
    }

    public boolean isOpen(Character c){
        char[] cAr = new char[] {'{', '(', '['};
        for(char ch : cAr) {
            if (ch == c){
                return true;
            }
        }
        return false;
    }
    public boolean isClosed(Character c){
        char[] cAr = new char[] {'}', ')', ']'};
        for(char ch : cAr) {
            if (ch == c){
                return true;
            }
        }
        return false;
    }


}
